sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/SessionReceiverManager.cs

```diff
@@ -0,0 +1,319 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Messaging.ServiceBus.Diagnostics;
+
+namespace Azure.Messaging.ServiceBus
+{
+    /// <summary>
+    /// Represents a thread-safe abstraction around a single session receiver that threads
+    /// spawned by the ServiceBusProcessor use to receive and process messages. Depending on how the
+    /// MaxConcurrentCalls and SessionIds options are configured, there may be multiple threads using the same
+    /// SessionReceiverManager (i.e. if MaxConcurrentCalls is greater than the number of specified sessions).
+    /// The manager will delegate to the user provided callbacks and handle automatic locking of sessions.
+    /// The receiver instance will only be closed when no other threads are using it, or when the user has
+    /// called StopProcessingAsync.
+    /// </summary>
+    internal class SessionReceiverManager : ReceiverManager
+    {
+        private int _threadCount = 0;
+        private readonly string _sessionId;
+        private readonly Func<ProcessSessionEventArgs, Task> _sessionInitHandler;
+        private readonly Func<ProcessSessionEventArgs, Task> _sessionCloseHandler;
+        private Func<ProcessSessionMessageEventArgs, Task> _messageHandler;
+        private readonly SemaphoreSlim _concurrentAcceptSessionsSemaphore;
+        private ServiceBusSessionReceiver _receiver;
+        protected override ServiceBusReceiver Receiver => _receiver;
+
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
+
+        public SessionReceiverManager(
+            ServiceBusConnection connection,
+            string fullyQualifiedNamespace,
+            string entityPath,
+            string identifier,
+            string sessionId,
+            ServiceBusProcessorOptions processorOptions,
+            Func<ProcessSessionEventArgs, Task> sessionInitHandler,
+            Func<ProcessSessionEventArgs, Task> sessionCloseHandler,
+            Func<ProcessSessionMessageEventArgs, Task> messageHandler,
+            Func<ProcessErrorEventArgs, Task> errorHandler,
+            SemaphoreSlim concurrentCallsSemaphore,
+            SemaphoreSlim concurrentAcceptSessionsSemaphore)
+            : base(connection, fullyQualifiedNamespace, entityPath, identifier, processorOptions, default, errorHandler, concurrentCallsSemaphore)
+        {
+            _sessionId = sessionId;
+            _sessionInitHandler = sessionInitHandler;
+            _sessionCloseHandler = sessionCloseHandler;
+            _messageHandler = messageHandler;
+            _concurrentAcceptSessionsSemaphore = concurrentAcceptSessionsSemaphore;
+        }
+
+        private async Task EnsureReceiverCreated(CancellationToken cancellationToken)
```

Comment:
Why do  you need both this and CreateAndInitializeSessionReceiver?

Reply:
Eh it is just an attempt to make the code more readable.